### DISCUSSION TITLE
error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools"

### DISCUSSION THREAD

**Discussion by bmartins-EUDA** [MaartenGr/BERTopic#2183] [General]
Dear community,

I have followed plenty of articles (see below) on solving this issue, which seems to have been reported many times in the past, but unsuccessfully.
https://www.scivision.dev/python-windows-visual-c-14-required/
https://github.com/bycloudai/InstallVSBuildToolsWindows?tab=readme-ov-file

Build tools are install using the following command:
`vs_buildtools.exe --quiet --wait --norestart --nocache --includeRecommended --includeOptional --add Microsoft.VisualStudio.Component.Windows10SDK.20348 --add Microsoft.VisualStudio.Component.VC.14.30.17.0.x86.x64 --add Microsoft.VisualStudio.Component.Windows11SDK.26100 --add Microsoft.VisualStudio.Workload.VCTools`

The full log is here:
> PS C:\WINDOWS\system32> pip install bertopic
> Collecting bertopic
>   Using cached bertopic-0.16.4-py3-none-any.whl.metadata (23 kB)
> Collecting hdbscan>=0.8.29 (from bertopic)
>   Using cached hdbscan-0.8.39.tar.gz (5.8 MB)
>   Installing build dependencies ... done
>   Getting requirements to build wheel ... done
>   Preparing metadata (pyproject.toml) ... done
> Collecting numpy>=1.20.0 (from bertopic)
>   Using cached numpy-2.1.2-cp312-cp312-win_amd64.whl.metadata (59 kB)
> Collecting pandas>=1.1.5 (from bertopic)
>   Using cached pandas-2.2.3-cp312-cp312-win_amd64.whl.metadata (19 kB)
> Collecting plotly>=4.7.0 (from bertopic)
>   Using cached plotly-5.24.1-py3-none-any.whl.metadata (7.3 kB)
> Collecting scikit-learn>=0.22.2.post1 (from bertopic)
>   Using cached scikit_learn-1.5.2-cp312-cp312-win_amd64.whl.metadata (13 kB)
> Collecting sentence-transformers>=0.4.1 (from bertopic)
>   Using cached sentence_transformers-3.2.0-py3-none-any.whl.metadata (10 kB)
> Collecting tqdm>=4.41.1 (from bertopic)
>   Using cached tqdm-4.66.5-py3-none-any.whl.metadata (57 kB)
> Collecting umap-learn>=0.5.0 (from bertopic)
>   Using cached umap_learn-0.5.6-py3-none-any.whl.metadata (21 kB)
> Collecting scipy>=1.0 (from hdbscan>=0.8.29->bertopic)
>   Using cached scipy-1.14.1-cp312-cp312-win_amd64.whl.metadata (60 kB)
> Collecting joblib>=1.0 (from hdbscan>=0.8.29->bertopic)
>   Using cached joblib-1.4.2-py3-none-any.whl.metadata (5.4 kB)
> Collecting python-dateutil>=2.8.2 (from pandas>=1.1.5->bertopic)
>   Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
> Collecting pytz>=2020.1 (from pandas>=1.1.5->bertopic)
>   Using cached pytz-2024.2-py2.py3-none-any.whl.metadata (22 kB)
> Collecting tzdata>=2022.7 (from pandas>=1.1.5->bertopic)
>   Using cached tzdata-2024.2-py2.py3-none-any.whl.metadata (1.4 kB)
> Collecting tenacity>=6.2.0 (from plotly>=4.7.0->bertopic)
>   Using cached tenacity-9.0.0-py3-none-any.whl.metadata (1.2 kB)
> Collecting packaging (from plotly>=4.7.0->bertopic)
>   Using cached packaging-24.1-py3-none-any.whl.metadata (3.2 kB)
> Collecting threadpoolctl>=3.1.0 (from scikit-learn>=0.22.2.post1->bertopic)
>   Using cached threadpoolctl-3.5.0-py3-none-any.whl.metadata (13 kB)
> Collecting transformers<5.0.0,>=4.41.0 (from sentence-transformers>=0.4.1->bertopic)
>   Using cached transformers-4.45.2-py3-none-any.whl.metadata (44 kB)
> Collecting torch>=1.11.0 (from sentence-transformers>=0.4.1->bertopic)
>   Using cached torch-2.4.1-cp312-cp312-win_amd64.whl.metadata (27 kB)
> Collecting huggingface-hub>=0.20.0 (from sentence-transformers>=0.4.1->bertopic)
>   Using cached huggingface_hub-0.25.2-py3-none-any.whl.metadata (13 kB)
> Collecting Pillow (from sentence-transformers>=0.4.1->bertopic)
>   Using cached pillow-10.4.0-cp312-cp312-win_amd64.whl.metadata (9.3 kB)
> Collecting colorama (from tqdm>=4.41.1->bertopic)
>   Using cached colorama-0.4.6-py2.py3-none-any.whl.metadata (17 kB)
> Collecting numba>=0.51.2 (from umap-learn>=0.5.0->bertopic)
>   Using cached numba-0.60.0-cp312-cp312-win_amd64.whl.metadata (2.8 kB)
> Collecting pynndescent>=0.5 (from umap-learn>=0.5.0->bertopic)
>   Using cached pynndescent-0.5.13-py3-none-any.whl.metadata (6.8 kB)
> Collecting filelock (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached filelock-3.16.1-py3-none-any.whl.metadata (2.9 kB)
> Collecting fsspec>=2023.5.0 (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached fsspec-2024.9.0-py3-none-any.whl.metadata (11 kB)
> Collecting pyyaml>=5.1 (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached PyYAML-6.0.2-cp312-cp312-win_amd64.whl.metadata (2.1 kB)
> Collecting requests (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)
> Collecting typing-extensions>=3.7.4.3 (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)
> Collecting llvmlite<0.44,>=0.43.0dev0 (from numba>=0.51.2->umap-learn>=0.5.0->bertopic)
>   Using cached llvmlite-0.43.0-cp312-cp312-win_amd64.whl.metadata (4.9 kB)
> Collecting numpy>=1.20.0 (from bertopic)
>   Using cached numpy-2.0.2-cp312-cp312-win_amd64.whl.metadata (59 kB)
> Collecting six>=1.5 (from python-dateutil>=2.8.2->pandas>=1.1.5->bertopic)
>   Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
> Collecting sympy (from torch>=1.11.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached sympy-1.13.3-py3-none-any.whl.metadata (12 kB)
> Collecting networkx (from torch>=1.11.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached networkx-3.4.1-py3-none-any.whl.metadata (6.3 kB)
> Collecting jinja2 (from torch>=1.11.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached jinja2-3.1.4-py3-none-any.whl.metadata (2.6 kB)
> Requirement already satisfied: setuptools in c:\users\martibr\appdata\local\programs\python\python312\lib\site-packages (from torch>=1.11.0->sentence-transformers>=0.4.1->bertopic) (75.1.0)
> Collecting regex!=2019.12.17 (from transformers<5.0.0,>=4.41.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached regex-2024.9.11-cp312-cp312-win_amd64.whl.metadata (41 kB)
> Collecting safetensors>=0.4.1 (from transformers<5.0.0,>=4.41.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached safetensors-0.4.5-cp312-none-win_amd64.whl.metadata (3.9 kB)
> Collecting tokenizers<0.21,>=0.20 (from transformers<5.0.0,>=4.41.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached tokenizers-0.20.1-cp312-none-win_amd64.whl.metadata (6.9 kB)
> Collecting MarkupSafe>=2.0 (from jinja2->torch>=1.11.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached MarkupSafe-3.0.1-cp312-cp312-win_amd64.whl.metadata (4.1 kB)
> Collecting charset-normalizer<4,>=2 (from requests->huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached charset_normalizer-3.4.0-cp312-cp312-win_amd64.whl.metadata (34 kB)
> Collecting idna<4,>=2.5 (from requests->huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached idna-3.10-py3-none-any.whl.metadata (10 kB)
> Collecting urllib3<3,>=1.21.1 (from requests->huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached urllib3-2.2.3-py3-none-any.whl.metadata (6.5 kB)
> Collecting certifi>=2017.4.17 (from requests->huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached certifi-2024.8.30-py3-none-any.whl.metadata (2.2 kB)
> Collecting mpmath<1.4,>=1.1.0 (from sympy->torch>=1.11.0->sentence-transformers>=0.4.1->bertopic)
>   Using cached mpmath-1.3.0-py3-none-any.whl.metadata (8.6 kB)
> Using cached bertopic-0.16.4-py3-none-any.whl (143 kB)
> Using cached pandas-2.2.3-cp312-cp312-win_amd64.whl (11.5 MB)
> Using cached plotly-5.24.1-py3-none-any.whl (19.1 MB)
> Using cached scikit_learn-1.5.2-cp312-cp312-win_amd64.whl (11.0 MB)
> Using cached sentence_transformers-3.2.0-py3-none-any.whl (255 kB)
> Using cached tqdm-4.66.5-py3-none-any.whl (78 kB)
> Using cached umap_learn-0.5.6-py3-none-any.whl (85 kB)
> Using cached huggingface_hub-0.25.2-py3-none-any.whl (436 kB)
> Using cached joblib-1.4.2-py3-none-any.whl (301 kB)
> Using cached numba-0.60.0-cp312-cp312-win_amd64.whl (2.7 MB)
> Using cached numpy-2.0.2-cp312-cp312-win_amd64.whl (15.6 MB)
> Using cached packaging-24.1-py3-none-any.whl (53 kB)
> Using cached pynndescent-0.5.13-py3-none-any.whl (56 kB)
> Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
> Using cached pytz-2024.2-py2.py3-none-any.whl (508 kB)
> Using cached scipy-1.14.1-cp312-cp312-win_amd64.whl (44.5 MB)
> Using cached tenacity-9.0.0-py3-none-any.whl (28 kB)
> Using cached threadpoolctl-3.5.0-py3-none-any.whl (18 kB)
> Using cached torch-2.4.1-cp312-cp312-win_amd64.whl (199.4 MB)
> Using cached transformers-4.45.2-py3-none-any.whl (9.9 MB)
> Using cached tzdata-2024.2-py2.py3-none-any.whl (346 kB)
> Using cached colorama-0.4.6-py2.py3-none-any.whl (25 kB)
> Using cached pillow-10.4.0-cp312-cp312-win_amd64.whl (2.6 MB)
> Using cached fsspec-2024.9.0-py3-none-any.whl (179 kB)
> Using cached llvmlite-0.43.0-cp312-cp312-win_amd64.whl (28.1 MB)
> Using cached PyYAML-6.0.2-cp312-cp312-win_amd64.whl (156 kB)
> Using cached regex-2024.9.11-cp312-cp312-win_amd64.whl (273 kB)
> Using cached safetensors-0.4.5-cp312-none-win_amd64.whl (286 kB)
> Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
> Using cached tokenizers-0.20.1-cp312-none-win_amd64.whl (2.4 MB)
> Using cached typing_extensions-4.12.2-py3-none-any.whl (37 kB)
> Using cached filelock-3.16.1-py3-none-any.whl (16 kB)
> Using cached jinja2-3.1.4-py3-none-any.whl (133 kB)
> Using cached networkx-3.4.1-py3-none-any.whl (1.7 MB)
> Using cached requests-2.32.3-py3-none-any.whl (64 kB)
> Using cached sympy-1.13.3-py3-none-any.whl (6.2 MB)
> Using cached certifi-2024.8.30-py3-none-any.whl (167 kB)
> Using cached charset_normalizer-3.4.0-cp312-cp312-win_amd64.whl (102 kB)
> Using cached idna-3.10-py3-none-any.whl (70 kB)
> Using cached MarkupSafe-3.0.1-cp312-cp312-win_amd64.whl (15 kB)
> Using cached mpmath-1.3.0-py3-none-any.whl (536 kB)
> Using cached urllib3-2.2.3-py3-none-any.whl (126 kB)
> Building wheels for collected packages: hdbscan
>   Building wheel for hdbscan (pyproject.toml) ... error
>   error: subprocess-exited-with-error
> 
>   × Building wheel for hdbscan (pyproject.toml) did not run successfully.
>   │ exit code: 1
>   ╰─> [30 lines of output]
>       C:\Users\martibr\AppData\Local\Temp\pip-build-env-m1zl9pfy\overlay\Lib\site-packages\setuptools\_distutils\dist.py:261: UserWarning: Unknown distribution option: 'test_suite'
>         warnings.warn(msg)
>       C:\Users\martibr\AppData\Local\Temp\pip-build-env-m1zl9pfy\overlay\Lib\site-packages\setuptools\_distutils\dist.py:261: UserWarning: Unknown distribution option: 'tests_require'
>         warnings.warn(msg)
>       running bdist_wheel
>       running build
>       running build_py
>       creating build\lib.win-amd64-cpython-312\hdbscan
>       copying hdbscan\branches.py -> build\lib.win-amd64-cpython-312\hdbscan
>       copying hdbscan\flat.py -> build\lib.win-amd64-cpython-312\hdbscan
>       copying hdbscan\hdbscan_.py -> build\lib.win-amd64-cpython-312\hdbscan
>       copying hdbscan\plots.py -> build\lib.win-amd64-cpython-312\hdbscan
>       copying hdbscan\prediction.py -> build\lib.win-amd64-cpython-312\hdbscan
>       copying hdbscan\robust_single_linkage_.py -> build\lib.win-amd64-cpython-312\hdbscan
>       copying hdbscan\validity.py -> build\lib.win-amd64-cpython-312\hdbscan
>       copying hdbscan\__init__.py -> build\lib.win-amd64-cpython-312\hdbscan
>       creating build\lib.win-amd64-cpython-312\hdbscan\tests
>       copying hdbscan\tests\test_branches.py -> build\lib.win-amd64-cpython-312\hdbscan\tests
>       copying hdbscan\tests\test_flat.py -> build\lib.win-amd64-cpython-312\hdbscan\tests
>       copying hdbscan\tests\test_hdbscan.py -> build\lib.win-amd64-cpython-312\hdbscan\tests
>       copying hdbscan\tests\test_prediction_utils.py -> build\lib.win-amd64-cpython-312\hdbscan\tests
>       copying hdbscan\tests\test_rsl.py -> build\lib.win-amd64-cpython-312\hdbscan\tests
>       copying hdbscan\tests\__init__.py -> build\lib.win-amd64-cpython-312\hdbscan\tests
>       running build_ext
>       C:\Users\martibr\AppData\Local\Temp\pip-build-env-m1zl9pfy\overlay\Lib\site-packages\Cython\Compiler\Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: C:\Users\martibr\AppData\Local\Temp\pip-install-dchstsvm\hdbscan_156cd2b1a0cd4075af2c888c81a9bbdc\hdbscan\_hdbscan_tree.pyx
>         tree = Parsing.p_module(s, pxd, full_module_name)
>       Compiling hdbscan/_hdbscan_tree.pyx because it changed.
>       [1/1] Cythonizing hdbscan/_hdbscan_tree.pyx
>       building 'hdbscan._hdbscan_tree' extension
>       error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
>   ERROR: Failed building wheel for hdbscan
> Failed to build hdbscan
> ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (hdbscan)

Is there any special trick to make this work?

Thanks for your help!

**Comment by MaartenGr** [comment]
Thanks for reaching out. There are a bunch of issues already discussing this that might give you additional help. Other than that, I did create a small guide discussing this for https://llm-book.com that you can find here: https://github.com/HandsOnLLM/Hands-On-Large-Language-Models/blob/main/.setup/conda/common_issues.md

**Reply by bmartins-EUDA** [reply]
That helped - thanks a million!

This is the command line version of the solution, which worked.
`vs_buildtools.exe --quiet --wait --norestart --nocache --includeRecommended --includeOptional --add Microsoft.VisualStudio.Component.Windows11SDK.22621 --add Microsoft.VisualStudio.Component.VC.Tools.x86.x64 --add Microsoft.VisualStudio.Component.VC.CMake.Project --add Microsoft.VisualStudio.Workload.VCTools --add Microsoft.VisualStudio.Component.TestTools.BuildTools`
